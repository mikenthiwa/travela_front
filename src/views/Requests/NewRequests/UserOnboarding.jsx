import React from 'react';
import { connect } from 'react-redux';
import Base from '../../Base';
import { UserOnboarding } from '../../../components/Forms';
import updateUserProfile from '../../../redux/actionCreator/userProfileActions';
import { getAllDepartment } from '../../../redux/actionCreator/userActions';
import { fetchRoleUsers } from '../../../redux/actionCreator/roleActions';
import { getOccupation } from '../../../redux/actionCreator/occupationActions';
import Preloader from '../../../components/Preloader/Preloader';
import RequestUtils from '../../../helper/request/RequestUtils';
import '../Requests.scss';

export const UserOnboardingRequestPage = (editing = false) => {
  class ConfirmUserDetails extends Base {
    constructor(props) {
      super(props);
      const {location} = this.props;
      this.state = {
        url: location.search,
        department: '',
      };
    }

    componentDidMount() {
      const {
        fetchRoleUsers, getAllDepartment,
        match, getOccupation
      } = this.props;
      const {params: {request_id}} = match;
      getOccupation();
      getAllDepartment();
      fetchRoleUsers(53019);
    }

    render() {
      const { updateUserProfile, userData, user,
        loading, errors, roleUsers, requestOnEdit, occupations,
        creatingRequest, history, departments,
        fetchingRequest, editingRequest,
      } = this.props;
      const { url } = this.state;
      const manager = requestOnEdit && requestOnEdit.manager;
      const managerName = RequestUtils.getManagerNameOrId(roleUsers, manager);
      if(requestOnEdit) requestOnEdit.manager = managerName;
      return (
        fetchingRequest || roleUsers.length === 0 || !userData.result ? (
          <div className="request-page__preloader">
            <Preloader />
          </div>
        ):
          (
            <UserOnboarding
              updateUserProfile={updateUserProfile} user={user} errors={errors}
              userData={userData && userData.result} occupations={occupations}
              loading={loading} managers={roleUsers} departments={departments}
              creatingRequest={creatingRequest}
              requestOnEdit={requestOnEdit}
              history={history}
              editing={editing} isEditing={editingRequest}
            />)
      );
    }
  }
  return ConfirmUserDetails;
};

const mapStateToProps = ({
  requests, user, role,  occupations,
}) => ({
  ...requests,
  ...role,
  ...occupations,

  isFetching: requests.isLoading,
  userData: user.getUserData,
  department: user.currentUser.department,
});

const actions = () => ({
  updateUserProfile,
  fetchRoleUsers,
  getOccupation,
  getAllDepartment
});

export default (editing = true) => {
  return connect(mapStateToProps, actions())(UserOnboardingRequestPage(editing));
};
