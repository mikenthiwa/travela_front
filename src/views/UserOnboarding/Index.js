import React, { Component } from 'react';
import { PropTypes } from 'prop-types';
import './userOnboardingForm.scss';
import OnBoardingTabHead from '../../components/RequestTab/OnboardingTabHead';
import ConfirmPersonalDetailsFieldset from '../../components/Forms/NewRequestForm/FormFieldsets/ConfirmPersonalDetails';
import travelStipendHelper from '../../helper/request/RequestUtils';
import FormContext from '../../components/Forms/FormsAPI/FormContext/FormContext';
import getDefaultBlanksValidatorFor from '../../components/Forms/FormsAPI/formValidator/confirmUserDetails';
import WelcomePage from '../../components/WelcomePage';
import UploadPassportPage from '../../components/PassportUpload/UploadPassport';

class UserOnboarding extends Component {
  constructor(props) {
    super(props);
    this.setUp();
    this.state = {
      ...this.defaultState,
      currentPage: 1,
      currentTab: 1,
      currentOrigin: 0,
    };
    this.validate = getDefaultBlanksValidatorFor(this);
  }

  componentDidMount() {
    const { managers } = this.props;
    const { values } = this.state;
    const managerChoices = managers.map(manager => manager.fullName);
    values.manager !== ''
      && managerChoices === [];

  }

  setUp = () => {
    const { managers } = this.props;
    const { name, gender, department, role, manager, location } = this.getPersonalDetails();
    const managerName = travelStipendHelper.getManagerNameOrId(managers, manager);
    this.defaultState = {
      values: {
        name: name,
        gender,
        department,
        role,
        location,
        manager: managerName,
      },
      errors: {},
      hasBlankFields: true,
      collapse: false,
      parentIds: 1,
      steps: [
        {
          id: 1,
          position: 1,
          name: 'Personal Information',
        },
        {
          id: 2,
          position: 2,
          name: 'Travel Document',
        },
      ],
    };
  };
  getPersonalDetails = (editing, detailsSource) => {
    const { userData, userDataUpdate: { result } } = this.props;
    const personalDetails = {};
    const personalDetailsAttributes = [
      'name',
      'gender',
      'department',
      'role',
      'manager',
      'location'
    ];
    const userGender = result ? result.gender : userData.gender;
    personalDetailsAttributes.map(attrb => {
      if (userData) {
        if (editing)
          return (personalDetails[attrb] = detailsSource[attrb]);
        userData.name = userData.passportName;
        userData.role = userData.occupation;
        userData.gender = userGender;
        let value = userData[attrb];
        value = !/^null|undefined$/.test(value) ? value : '';
        return (personalDetails[attrb] = value);
      }
    });
    return personalDetails;
  };
  handleSubmit = () => {
    const { updateUserProfile, user, managers } = this.props;
    const userId = user.UserInfo.id;
    const { values } = this.state;
    if (this.validate) {
      const managerId = travelStipendHelper.getManagerNameOrId(managers, values.manager);
      const data = {
        fullName: values.name,
        passportName: values.name,
        occupation: values.role,
        manager: managerId,
        location: values.location,
        gender: values.gender,
        department: values.department,
      };
      updateUserProfile(data, userId, true);
      this.setState({ hasBlankFields: true });
      localStorage.setItem('location', values.location);
    }

  }

  changeState = (InputState, ErrorMessageState, errorMessage, value) => {
    const { occupations, managers } = this.props;
    const getUser = this.getPersonalDetails();
    const userDepartment = Object.keys(getUser).map((key) => {
      return [(key), getUser[key]];
    });
    const mainDepartment = userDepartment.filter(department => department.includes('department'));
    const mainLocation = userDepartment.filter(location => location.includes('location'));
    this.setState((prevState) => {
      const newState = { ...prevState.values, ...InputState };
      return { ...prevState, values: { ...newState } };
    });
    const departmentSuggestionChoices = 'manager' in InputState ? managers.map(manager => manager.fullName) :
      mainDepartment.map(department => department[1]);
    const locationSuggestionChoices  = 'manager' in InputState ? managers.map(manager => manager.fullName) :
      mainLocation.map(location => location[1]);
    const NameManagerSuggestionChoices = 'manager' in InputState ? managers.map(manager => manager.fullName) :
      occupations.map(occupation => occupation.occupationName);
    if ((NameManagerSuggestionChoices.indexOf(value) === -1) && (locationSuggestionChoices.indexOf(value) === -1) &&
      (departmentSuggestionChoices.indexOf(value) === -1)) return this.setError(errorMessage);
    return this.setState((prevState) => {
      const newError = { ...prevState.errors, ...ErrorMessageState };
      return { ...prevState, errors: { ...newError } };
    });
  }
  onChangeAutoSuggestion = (type, value) => {
    let errorMessage = { manager: 'No manager with the name exists' };
    let rolesErrorMessage = { role: 'No role with the name exists' };
    let departmentErrorMessage = { department: 'No department with the name exists' };
    let locationErrorMessage = { location: 'No location with the name exists' };
    const managerInputState = { 'manager': value };
    const managerErrorMessageState = { 'manager': '' };
    const occupationErrorMessageState = { 'role': '' };
    const occupationInputState = { 'role': value };
    const departmentInputState = { 'department': value };
    const departmentErrorMessageState = { 'department': '' };
    const locationInputState = { 'location': value };
    const locationErrorMessageState = { 'location': '' };
    switch (type) {
    case 'manager':
      this.changeState(managerInputState, managerErrorMessageState, errorMessage, value);
      break;
    case 'role':
      this.changeState(occupationInputState, occupationErrorMessageState, rolesErrorMessage, value);
      break;
    case 'department':
      this.changeState(departmentInputState, departmentErrorMessageState, departmentErrorMessage, value);
      break;
    case 'location':
      this.changeState(locationInputState, locationErrorMessageState, locationErrorMessage, value);
      break;
    default:
      null;
    }
  };
  setError = (errorMessage) => {
    return this.setState((prevState) => {
      const newError = {
        ...prevState.errors,
        ...errorMessage
      };
      return { ...prevState, errors: { ...newError } };
    });
  };


  nextStep = (e) => {
    e.preventDefault();
    const { steps, currentPage, currentTab } = this.state;
    if (currentPage !== 2) {
      this.setState({
        currentPage: currentPage + 1
      });
    } else {
      if (currentTab !== 2) {
        this.handleSubmit();
        this.setState({
          currentTab: currentTab + 1,
          currentPage: currentPage + 1
        });
      }
    }
  };

  previousStep = () => {
    const { currentPage , currentTab } = this.state;
    if (currentPage === 3) {
      this.setState({ currentPage: currentPage - 1, currentTab: currentTab -1 });
    }
  };

  renderPersonalDetailsFieldset = () => {
    const { collapse, title, position, line, values, errors } = this.state;
    const { managers, occupations, creatingRequest, departments, userData: { department } } = this.props;

    const existingDepartment = department;

    const getUserDetails = this.getPersonalDetails();
    return (
      <div>
        <ConfirmPersonalDetailsFieldset
          departments={departments}
          existingDepartment={existingDepartment}
          values={values}
          savePersonalDetails={this.savePersonalDetails}
          onChangeAutoSuggestion={this.onChangeAutoSuggestion}
          collapsible={this.collapsible}
          collapse={collapse}
          getUserDetails={getUserDetails}
          managers={managers}
          occupations={occupations}
          value="100%"
          hasBlankFields={
            !!errors.manager || !!errors.role || !!errors.department || !!errors.name || !!errors.location
          }
          loading={creatingRequest}
          send="Next"
          nextStep={this.nextStep}
        />
      </div>

    );
  };

  savePersonalDetails(personalDetails) {
    Object.keys(personalDetails).forEach(key => {
      localStorage.setItem(key, personalDetails[key]);
    });
  }

  renderWelcomePage = () => {
    const { errors } = this.state;
    const { creatingRequest, userData } = this.props;
    return (
      <div className="user_onboarding">
        <WelcomePage
          nextStep={this.nextStep}
          hasBlankFields={
            !!errors.manager
          }
          send="Get Started"
          loading={creatingRequest}
          fullName={userData.fullName}
        />
      </div>
    );
  }
  renderUploadPassportPage = () => {
    const { history } = this.props;
    return (
      <div>
        <UploadPassportPage history={history} />

        <button
          type="button"
          className="back-btn-travel-document back-btn--active"
          onClick={this.previousStep}
        >
       Back
        </button>

      </div>
    );
  }

  renderForm = () => {
    const {
      errors, values,
      steps, currentPage, currentTab,
    } = this.state;
    const {
      editing, history,
    } = this.props;
    return (
      <div className="width-911">
        {currentPage !== 1 ? (
          <OnBoardingTabHead
            steps={steps} currentTab={currentTab}
            currentPage={currentPage} editing={editing} history={history} />
        ) : null}
        <FormContext
          steps={steps}
          editing={editing} history={history}
          currentPage={currentPage}
          currentTab={currentTab}
          targetForm={this}
          values={values}
          errors={errors}
          validatorName="validate">
          <form onSubmit={() => this.handleSubmit()} className="new-request">
            { currentPage === 1 && this.renderWelcomePage()}
            { currentPage === 2 && this.renderPersonalDetailsFieldset()}
            { currentPage === 3 &&this.renderUploadPassportPage()}
          </form>
        </FormContext>
      </div>

    );

  };
  render() {
    return <div>{this.renderForm()}</div>;
  }
}

UserOnboarding.propTypes = {
  user: PropTypes.object.isRequired,
  updateUserProfile: PropTypes.func.isRequired,
  userData: PropTypes.object,
  managers: PropTypes.array,
  departments: PropTypes.array,
  occupations: PropTypes.array,
  creatingRequest: PropTypes.bool,
  editing: PropTypes.bool,
  userDataUpdate: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.object
  ]),
  history: PropTypes.object,
};

UserOnboarding.defaultProps = {
  creatingRequest: false,
  editing: false,
  managers: [],
  departments: [],
  occupations: [],
  userData: {},
  userDataUpdate: [],
  history: {}
};

export default UserOnboarding;
