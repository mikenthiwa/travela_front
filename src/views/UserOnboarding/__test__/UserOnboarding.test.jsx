import React from 'react';
import sinon from 'sinon';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import createSagaMiddleware from 'redux-saga';
import { shallow } from 'enzyme';
import UserOnboarding from '../Index';

describe('<UserOnboarding />', () => {
  let wrapper, onSubmit, onClick;
  onSubmit = jest.fn();
  onClick = jest.fn();


  const props = {
    loading: false,
    hasBlankFields: false,
    user: {
      UserInfo: {
        id: '-LJNw1BsT0LP_E4l2peP',
        name: 'Collins',
      }
    },
    userData: {
      id: '2',
      fullName: 'Collins Muru',
      name: 'Collins',
      email: 'collins.muru@andela.com',
      userId: '-LJNw1BsT0LP_E4l2peP',
      passportName: 'Collins',
      department: 'Talent & Development',
      occupation: 'Software Developer',
      manager: 'Collins',
      gender: 'Male',
      createdAt: '2018-09-14T12:48:11.266Z',
      updatedAt: '2018-09-16T07:53:48.835Z',
      roleId: 401938,
      location: 'Kigali'
    },


    userDataUpdate: [],
    requestOnEdit: {
      id: '1',
      name: 'Seun Undefined',
      manager: 'Faluyi Seun',
      gender: 'Male',
      department: 'Talent & Development',
      role: 'Software Developer',
      status: 'Open',
      userId: 'lorem-ipsum',
      createdAt: '2018-09-26T15:15:49.808Z',
      updatedAt: '2018-09-26T15:15:49.808Z',
      
    },
    google: {},
    getUserData: jest.fn(() => {
    }),
    updateUserProfile: jest.fn(() => {
    }),
    getPersonalDetails: jest.fn(() => {
    }),
    onChangeAutoSuggestion: jest.fn(),
    choices: ['director', 'chef'],
    managers: [{
      fullName: 'Test User',
      email: 'test.user@andela.com'
    },
    {
      fullName: 'Samuel Kubai',
      email: 'samuel@andela.com'
    }],
    history: {
      push: jest.fn()
    },
    handleSubmit:jest.fn()
  };
  const user = localStorage.getItem('name');
  const location = localStorage.getItem('location');
  const gender = localStorage.getItem('gender');
  const department = localStorage.getItem('department');
  const role = localStorage.getItem('role');
  const manager = localStorage.getItem('manager');

  const defaultState = {
    values: {
      name: !(/^null|undefined$/).test(user) ? user : '', 
      gender: !(/^null|undefined$/).test(gender) ? gender : '',
      department: !(/^null|undefined$/).test(department) ? department : '',
      role: !(/^null|undefined$/).test(role) ? role : '',
      manager: !(/^null|undefined$/).test(manager) ? manager : '',
      location: !(/^null|undefined$/).test(location) ? location : '',
    },
    errors: {},
    hasBlankFields: true,
    collapse: false,
    parentIds: 1,
    steps: [
      { id: 1,  name: 'Personal Information' },
      { id: 2,  name: 'Travel Document' },
    ],
    currentTab: 1,
    currentOrigin: 0,
    currentPage: 1,
  };

  beforeEach(() => {
    wrapper = mount(<UserOnboarding {...props} />);
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should profile information after update ', () => {
    const shallowWrapper = shallow(<UserOnboarding {...props} />);
    localStorage.setItem('checkBox', 'clicked');
    shallowWrapper.setState({
      values: {
        name: 'Moses Gitau', 
        gender: 'male',
        department: 'Talent Driven Development',
        role: 'Software Developer',
        manager: 'Samuel Kubai',
      },
      
    });
    const event = {
      preventDefault: jest.fn(),
    };
    sinon.spy(shallowWrapper.instance(), 'handleSubmit');
    shallowWrapper.instance().handleSubmit(event);
    expect(shallowWrapper.instance().handleSubmit.calledOnce).toEqual(true);
    expect(props.updateUserProfile).toHaveBeenCalledTimes(1);
  });
  it('Test HandleSubmit button ', () => {
    const spy = sinon.spy();
    const wrapper = shallow(<UserOnboarding {...props} handleSubmit={spy} />);
    wrapper.find('form').at(0).simulate('click');
    expect(spy.wrapper);
  });

  it('renders the personal details fieldset', () => {
    const personalDetails = wrapper.find('fieldset.personal-details');
    expect(personalDetails);
  });

  it('renders one buttons', () => {
    const buttons = wrapper.find('button');
    expect(buttons).toHaveLength(1);
  });
  it('Test the onclick button for the next step action', () => {
    const clickButton = wrapper.find('request-submit-area-details');

  });
  it('validates form before sending data', () => {
    const form = wrapper.find('form');
    form.simulate('submit');
    expect(onSubmit).toHaveBeenCalledTimes(0);
  });

  it('should test onChangeManager()', () => {
    const shallowWrapper = mount(<UserOnboarding {...props} />);
    localStorage.setItem('checkBox', 'clicked');
    shallowWrapper.setState({
      values: {
        name: 'Akanmu Chris',
        gender: 'male',
        department: 'Success',
        role: 'Software Developer',
        manager: 'Samuel Kubai',
      },
    });
    const params = ['manager', 'David Ssalli'];
    expect(shallowWrapper.instance().onChangeAutoSuggestion(...params));
    const { manager } = shallowWrapper.state('values');
    expect(manager).toEqual('David Ssalli');

    const params2 = ['manager', 'will fail'];
    expect(shallowWrapper.instance().onChangeAutoSuggestion(...params2));
    const { manager: newManager } = shallowWrapper.state('values');
    const { manager: secondManagerError } = shallowWrapper.state('errors');
    expect(newManager).toEqual('will fail');
    expect(secondManagerError).toEqual('No manager with the name exists');
    
  });

  it('should test onChangeOccupation()', () => {
    const shallowWrapper = mount(<UserOnboarding {...props} />);
    localStorage.setItem('checkBox', 'clicked');
    shallowWrapper.setState({
      values: {
        name: 'Kevin Munene',
        gender: 'male',
        department: 'Success',
        role: 'Software Developer',
        manager: 'David Ssali',
      },
    });

    const params = ['role', 'Software Developers'];
    expect(shallowWrapper.instance().onChangeAutoSuggestion(...params));
    const { role } = shallowWrapper.state('values');
    expect(role).toEqual('Software Developers');

    const params2 = ['role', 'Journalist'];
    expect(shallowWrapper.instance().onChangeAutoSuggestion(...params2));
    const { role: newRole } = shallowWrapper.state('values');
    const { role: secondRoleError } = shallowWrapper.state('errors');
    expect(newRole).toEqual('Journalist');
    expect(secondRoleError).toEqual('No role with the name exists');
    
  });

  it('should test onChangeLocation()', () => {
    const shallowWrapper = mount(<UserOnboarding {...props} />);
    localStorage.setItem('checkBox', 'clicked');
    shallowWrapper.setState({
      values: {
        name: 'Kevin Munene',
        gender: 'male',
        department: 'Success',
        role: 'Software Developer',
        manager: 'David Ssali',
        location: 'Lagos'
      },
    });
    const params = ['location', 'Lagos'];
    expect(shallowWrapper.instance().onChangeAutoSuggestion(...params));
    const { location } = shallowWrapper.state('values');
    expect(location).toEqual('Lagos');

    const params2 = ['location', 'Aba'];
    expect(shallowWrapper.instance().onChangeAutoSuggestion(...params2));
    const { location: newLocation } = shallowWrapper.state('values');
    const { location: secondLocationError } = shallowWrapper.state('errors');
    expect(newLocation).toEqual('Aba');
    expect(secondLocationError).toEqual('No location with the name exists');
    
  });

  it('should test onChangeDeparment()', () => {
    const shallowWrapper = mount(<UserOnboarding {...props} />);
    localStorage.setItem('checkBox', 'clicked');
    shallowWrapper.setState({
      values: {
        name: 'Kevin Munene',
        gender: 'male',
        department: 'Success',
        role: 'Software Developer',
        manager: 'David Ssali',
        
      },
    });
    const params = ['department', 'Success'];
    expect(shallowWrapper.instance().onChangeAutoSuggestion(...params));
    const { department } = shallowWrapper.state('values');
    expect(department).toEqual('Success');

    const params2 = ['department', 'Actor'];
    expect(shallowWrapper.instance().onChangeAutoSuggestion(...params2));
    const { department: newDepartment } = shallowWrapper.state('values');
    const { department: secondDepartmentError } = shallowWrapper.state('errors');
    expect(newDepartment).toEqual('Actor');
    expect(secondDepartmentError).toEqual('No department with the name exists');
    
  });

  it('It should not return any onChangeAutoSuggestion', () => {
    const shallowWrapper = mount(<UserOnboarding {...props} />);
    localStorage.setItem('checkBox', 'clicked');
    shallowWrapper.setState({
      values: null
    });
    const params = [];
    expect(shallowWrapper.instance().onChangeAutoSuggestion(params));

    
  });
  it('should call localStorage when savePersonalDetails is called', () => {
    const wrapper = shallow(<UserOnboarding {...props} />);
    wrapper.instance().savePersonalDetails({ key: 'value' });
    expect(localStorage.getItem('key')).toEqual('value');
  });

  it('should change to the next step ', () => {
    const shallowWrapper = shallow(<UserOnboarding {...props} />);
    const event = {
      preventDefault: jest.fn(),
    };
    sinon.spy(shallowWrapper.instance(), 'nextStep');
    shallowWrapper.instance().nextStep(event);
    expect(shallowWrapper.instance().nextStep.calledOnce).toEqual(true);
  });
  it('should display next step on personal information  ', () => {
    const shallowWrapper = mount(<UserOnboarding {...props} />);
    shallowWrapper.setState({
      currentTab: 1
    });
    const event = {
      preventDefault: jest.fn(),
      target: {
        name: 'Personal Information',
      }
    };
    expect(shallowWrapper.state().currentTab).toEqual(1);
  });

  describe('Request Edit', () => {
    const sagaMiddleware = [createSagaMiddleware];
    const mockStore = configureStore(sagaMiddleware);
    const store = mockStore({
      requests: {
        requestOnEdit: props.requestOnEdit,
        requestData: {}
      },

    });
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <UserOnboarding {...props} editing />
        </MemoryRouter>
      </Provider>
    );

    it('should not submit data without validation', () => {
      const requestForm = wrapper.find('form');
      requestForm.simulate('submit');
      expect(onSubmit).toHaveBeenCalledTimes(0);
    });
    it('should render main div correctly', () => {
      const wrapper = shallow(<UserOnboarding />);
      const currentTab = 1;
      const defaultState = {
        currentPage: 1,
        currentTab: 1,
        currentOrigin: 0,
      };
      wrapper.instance().setState(defaultState);
      expect(wrapper.find('.user_onboarding').length).toBe(currentTab);
    });
    it('renders correctly', () => {
      const wrapper = shallow(<UserOnboarding />);
      expect(wrapper).toMatchSnapshot();
    });

    it('should to change to the next step ', () => {
      const shallowWrapper = shallow(<UserOnboarding />);
      const defaultState = {
        currentTab: 1,
        currentPage: 1,
        steps: '',
      };
      shallowWrapper.instance().setState(defaultState);
      const event = {
        preventDefault: jest.fn(),
      };
      sinon.spy(shallowWrapper.instance(), 'nextStep');
      shallowWrapper.instance().nextStep(event);
      expect(shallowWrapper.instance().nextStep.calledOnce).toEqual(true);
    });

    it('should submit user details when the next button is click ', () => {
      const spy = sinon.spy();
      const shallowWrapper = shallow(<UserOnboarding {...props} handleSubmit={spy} />);
      const defaultState = {
        currentTab: 1,
        currentPage: 2,
        steps: ''
      };
      shallowWrapper.instance().setState(defaultState);
      const event = {
        preventDefault: jest.fn(),
      };
      shallowWrapper.find('form').at(0).simulate('click');
      sinon.spy(shallowWrapper.instance(), 'nextStep');
      sinon.spy(shallowWrapper.instance(), 'handleSubmit');
      const res = shallowWrapper.instance().nextStep(event);
      expect(shallowWrapper.instance().nextStep.calledOnce).toEqual(true);
      expect(shallowWrapper.instance().handleSubmit.calledOnce).toEqual(true);
      expect(spy.wrapper);
    });

  });
  it('validates required field at initial state', () => {
    wrapper.instance().validate();
    expect(wrapper.state('errors').undefined).toEqual('This field is required');
  });
  
});
