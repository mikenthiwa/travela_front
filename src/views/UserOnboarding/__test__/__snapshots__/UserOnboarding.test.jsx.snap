// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserOnboarding /> Request Edit renders correctly 1`] = `
<div>
  <div
    className="width-911"
  >
    <FormContext
      currentPage={1}
      currentTab={1}
      editing={false}
      errors={Object {}}
      history={Object {}}
      steps={
        Array [
          Object {
            "id": 1,
            "name": "Personal Information",
            "position": 1,
          },
          Object {
            "id": 2,
            "name": "Travel Document",
            "position": 2,
          },
        ]
      }
      targetForm={
        UserOnboarding {
          "changeState": [Function],
          "context": Object {},
          "defaultState": Object {
            "collapse": false,
            "errors": Object {},
            "hasBlankFields": true,
            "parentIds": 1,
            "steps": Array [
              Object {
                "id": 1,
                "name": "Personal Information",
                "position": 1,
              },
              Object {
                "id": 2,
                "name": "Travel Document",
                "position": 2,
              },
            ],
            "values": Object {
              "department": "",
              "gender": "",
              "location": "",
              "manager": null,
              "name": "",
              "role": "",
            },
          },
          "getPersonalDetails": [Function],
          "handleSubmit": [Function],
          "nextStep": [Function],
          "onChangeAutoSuggestion": [Function],
          "props": Object {
            "creatingRequest": false,
            "editing": false,
            "history": Object {},
            "managers": Array [],
            "occupations": Array [],
            "userData": Object {
              "gender": undefined,
              "name": undefined,
              "role": undefined,
            },
            "userDataUpdate": Array [],
          },
          "refs": Object {},
          "renderForm": [Function],
          "renderPersonalDetailsFieldset": [Function],
          "renderWelcomePage": [Function],
          "setError": [Function],
          "setState": [Function],
          "setUp": [Function],
          "state": Object {
            "collapse": false,
            "currentOrigin": 0,
            "currentPage": 1,
            "currentTab": 1,
            "errors": Object {},
            "hasBlankFields": true,
            "parentIds": 1,
            "steps": Array [
              Object {
                "id": 1,
                "name": "Personal Information",
                "position": 1,
              },
              Object {
                "id": 2,
                "name": "Travel Document",
                "position": 2,
              },
            ],
            "values": Object {
              "department": "",
              "gender": "",
              "location": "",
              "manager": null,
              "name": "",
              "role": "",
            },
          },
          "updater": Updater {
            "_callbacks": Array [],
            "_renderer": ReactShallowRenderer {
              "_context": Object {},
              "_element": <UserOnboarding
                creatingRequest={false}
                editing={false}
                history={Object {}}
                managers={Array []}
                occupations={Array []}
                userData={
                        Object {
                                "gender": undefined,
                                "name": undefined,
                                "role": undefined,
                              }
                }
                userDataUpdate={Array []}
        />,
              "_forcedUpdate": false,
              "_instance": [Circular],
              "_newState": null,
              "_rendered": <div>
                <div
                        className="width-911"
                >
                        <FormContext
                                currentPage={1}
                                currentTab={1}
                                editing={false}
                                errors={Object {}}
                                history={Object {}}
                                steps={
                                        Array [
                                                Object {
                                                  "id": 1,
                                                  "name": "Personal Information",
                                                  "position": 1,
                                                },
                                                Object {
                                                  "id": 2,
                                                  "name": "Travel Document",
                                                  "position": 2,
                                                },
                                              ]
                                }
                                targetForm={[Circular]}
                                validatorName="validate"
                                values={
                                        Object {
                                                "department": "",
                                                "gender": "",
                                                "location": "",
                                                "manager": null,
                                                "name": "",
                                                "role": "",
                                              }
                                }
                        >
                                <form
                                        className="new-request"
                                        onSubmit={[Function]}
                                >
                                        <div
                                                className="user_onboarding"
                                        >
                                                <WelcomePage
                                                        fullName=""
                                                        hasBlankFields={false}
                                                        loading={false}
                                                        nextStep={[Function]}
                                                        send="Get Started"
                                                />
                                        </div>
                                </form>
                        </FormContext>
                </div>
        </div>,
              "_rendering": false,
              "_updater": [Circular],
            },
          },
          "validate": [Function],
          Symbol(enzyme.__setState__): [Function],
        }
      }
      validatorName="validate"
      values={
        Object {
          "department": "",
          "gender": "",
          "location": "",
          "manager": null,
          "name": "",
          "role": "",
        }
      }
    >
      <form
        className="new-request"
        onSubmit={[Function]}
      >
        <div
          className="user_onboarding"
        >
          <WelcomePage
            fullName=""
            hasBlankFields={false}
            loading={false}
            nextStep={[Function]}
            send="Get Started"
          />
        </div>
      </form>
    </FormContext>
  </div>
</div>
`;

exports[`<UserOnboarding /> renders correctly 1`] = `
<div>
  <div
    className="width-911"
  >
    <form
      className="new-request"
      key=".0"
      onSubmit={[Function]}
    >
      <div
        className="user_onboarding"
      >
        <div
          className="welcome"
        >
          <div
            className="heading"
          >
            Welcome to Travela
          </div>
          <img
            alt="welcome"
            className="welcome_image"
            src="welcome_page.svg"
          />
          <div
            className="message"
          >
            Hi Collins Muru this is your first time here & we're glad to have you on board. please click the button to begin
          </div>
          <button
            className="get_started"
            disabled={false}
            id="get-started"
            onClick={[Function]}
            type="submit"
          >
            <undefined />
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
